code/parachain/frame/assets-interface/src/lib.rs

```diff
@@ -0,0 +1,344 @@
+			ensure!(asset_admin == who, Error::<T>::NotAssetAdmin);
+			<T::AssetsRegistry as RemoteAssetRegistryMutate>::set_reserve_location(
+				asset_id, location,
+			)?;
+			Ok(())
+		}
+
+		#[pallet::call_index(4)]
+		#[pallet::weight(<T as Config>::WeightInfo::mint_cosmwasm())]
+		pub fn mint_cosmwasm(
+			origin: OriginFor<T>,
+			asset_id: T::AssetId,
+			dest: T::AccountId,
+			amount: T::Balance,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			let asset_admin =
+				AssetAdmin::<T>::try_get(asset_id).map_err(|_| Error::<T>::AssetNotFound)?;
+			ensure!(asset_admin == who, Error::<T>::NotAssetAdmin);
+			<Self as CosmWasmIssuance>::mint_into(asset_id, dest, amount)?;
+			Ok(())
+		}
+
+		#[pallet::call_index(5)]
+		#[pallet::weight(<T as Config>::WeightInfo::burn_cosmwasm())]
+		pub fn burn_cosmwasm(
+			origin: OriginFor<T>,
+			asset_id: T::AssetId,
+			dest: T::AccountId,
+			amount: T::Balance,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			let asset_admin =
+				AssetAdmin::<T>::try_get(asset_id).map_err(|_| Error::<T>::AssetNotFound)?;
+			ensure!(asset_admin == who, Error::<T>::NotAssetAdmin);
+			<Self as CosmWasmIssuance>::burn_from(asset_id, dest, amount)?;
+			Ok(())
+		}
+	}
+
+	impl<T: Config> PermissionlessAsset for Pallet<T> {
+		type AccountId = T::AccountId;
+		type AssetId = T::AssetId;
+		type AssetNativeLocation = T::Location;
+		type Balance = T::Balance;
+
+		fn register_asset(
+			account_id: Self::AccountId,
+			asset_id: Self::AssetId,
+			location: T::Location,
+			name: Option<BiBoundedAssetName>,
+			symbol: Option<BiBoundedAssetSymbol>,
+			decimals: u8,
+			is_cosmwasm: bool,
+		) -> DispatchResult {
+			ensure!(decimals > 0 && decimals < 30, Error::<T>::BadDecimals);
+
+			<T::AssetsRegistry as AssetExistentialDepositInspect>::existential_deposit(asset_id)?;
+
+			let asset_info: AssetInfo<T::Balance> = AssetInfo {
+				name,
+				symbol,
+				decimals: Some(decimals),
+				existential_deposit: T::Balance::from(1_u128),
+				ratio: None,
+			};
+			let fee_amount = AssetCreationFee::<T>::get().ok_or(Error::<T>::CreationFeeIsNotSet)?;
+			let result = T::Currency::transfer(
+				&account_id,
+				&T::TreasuryAccount::get(),
+				fee_amount,
+				KeepAlive,
+			);
+			ensure!(result.is_ok(), Error::<T>::TransferError);
+
+			<T::AssetsRegistry as RemoteAssetRegistryMutate>::register_asset(
+				asset_id,
+				Some(location),
+				asset_info,
+			)?;
+
+			AssetAdmin::<T>::insert(&asset_id, account_id);
+			if is_cosmwasm {
+				IsCosmWasmAsset::<T>::insert(&asset_id, ());
+			}
+
+			Ok(())
+		}
+	}
+	impl<T: Config> CosmWasmIssuance for Pallet<T> {
+		type AccountId = T::AccountId;
+		type AssetId = T::AssetId;
+		type Balance = T::Balance;
+		fn mint_into(
+			asset_id: Self::AssetId,
+			dest: Self::AccountId,
+			amount: Self::Balance,
+		) -> DispatchResult {
+			IsCosmWasmAsset::<T>::try_get(asset_id).map_err(|_| Error::<T>::NotCosmWasmAsset)?;
```

Comment:
IBC/XCM should mint too.

Reply:
i am not sure that this is true. the initial goal for permissionless assets was to be bridged to picasso. these assets shouldnt be minted anyone. all of the  assets are can be minted via sudo via Assets or AssetsTranasctorRouter. This mint/burn is only for cosmwasm contracts assets.

Reply:
what is the difference for cw contract and external transaction via pdjs?

Reply:
the way i see it is that any cosmwasm developer should be allowed to create their tokens if they need for any of their smart contracts. thus they should have more control over them. all other tokens we dont want to allow minting and burning except by sudo or governance

Reply:
think about cw security model. there is no difference amid pd.js tx and cw call.

Reply:
what more control means? terms are too vague.

Reply:
there is a difference. cw can be minted by any cw contract which basically means by anyone, all other assets are minted by particular pallet logic, we dont allow it to be minted by extrinsic if it isnt sudo. cw contracts can be minted by the asset admin in any way they want. if assets registry and assets are currently isolated, AR stores data, assets act as router. AR doesnt know about assets, assets dont have any storage. if we move logic of permissionless and cw assets into AR then assets need to know about who created asset so AR will need to modify assets storage. these 2 pallets are independent rn, we will make codependent on each other. i was trying to avoid adding storage to assets router and make AR independent of the router.

Reply:
owner into assets storage. trait with impl using ar and assers traits. no new pallet. thinkn about it.